zcash_primitives/src/transaction/builder.rs

```diff
@@ -306,14 +440,76 @@ impl<'a, P: consensus::Parameters, R: RngCore> Builder<'a, P, R> {
             .ok_or(Error::InvalidAmount)
     }
 
+    pub fn compute_balances<BalanceError>(
+        &self,
+        fee_strategy: &impl FeeStrategy<Error = BalanceError>,
+    ) -> Result<TransactionBalances, BalanceError>
+    where
+        Error: From<BalanceError>,
+    {
+        fee_strategy.compute_balances(
+            self.transparent_builder.inputs(),
+            self.transparent_builder.outputs(),
+            self.sapling_builder.inputs(),
+            self.sapling_builder.outputs(),
+        )
+    }
+
+    #[cfg(feature = "zfuture")]
+    pub fn compute_balances_zfuture<BalanceError>(
+        &self,
+        fee_strategy: &impl FutureFeeStrategy<Error = BalanceError>,
+    ) -> Result<TransactionBalances, BalanceError>
+    where
+        Error: From<BalanceError>,
+    {
+        fee_strategy.compute_balances_zfuture(
+            self.transparent_builder.inputs(),
+            self.transparent_builder.outputs(),
+            self.sapling_builder.inputs(),
+            self.sapling_builder.outputs(),
+            self.tze_builder.inputs(),
+            self.tze_builder.outputs(),
+        )
+    }
+
     /// Builds a transaction from the configured spends and outputs.
     ///
     /// Upon success, returns a tuple containing the final transaction, and the
     /// [`SaplingMetadata`] generated during the build process.
-    pub fn build(
-        mut self,
+    pub fn build<BalanceError>(
```

Comment:
- Document this change in the changelog.
- Similarly to above, this should not take the error as the generic:

```suggestion
    pub fn build<S: FeeStrategy>(
```